Python/compile.c

```diff
@@ -7921,81 +7935,94 @@ scan_block_for_local(int target, basicblock *b, bool unsafe_to_start,
         assert(instr->i_opcode != LOAD_CONST__LOAD_FAST);
         assert(instr->i_opcode != STORE_FAST__STORE_FAST);
         assert(instr->i_opcode != LOAD_FAST__LOAD_CONST);
-        if (unsafe && instr->i_except != NULL) {
-            MAYBE_PUSH(instr->i_except);
+        if (instr->i_except != NULL) {
+            maybe_push(instr->i_except, unsafe_mask, sp);
         }
-        if (instr->i_oparg != target) {
+        if (instr->i_oparg >= 64) {
             continue;
         }
+        assert(instr->i_oparg >= 0);
+        uint64_t bit = (uint64_t)1 << instr->i_oparg;
         switch (instr->i_opcode) {
-            case LOAD_FAST_CHECK:
-                // if this doesn't raise, then var is defined
-                unsafe = false;
+            case DELETE_FAST:
+                unsafe_mask |= bit;
                 break;
             case LOAD_FAST:
-                if (unsafe) {
+                // If this doesn't raise, then var is defined.
+                if (unsafe_mask & bit) {
                     instr->i_opcode = LOAD_FAST_CHECK;
                 }
-                unsafe = false;
+                unsafe_mask &= ~bit;
                 break;
-            case STORE_FAST:
-                unsafe = false;
+            case LOAD_FAST_CHECK:
+                unsafe_mask &= ~bit;
                 break;
-            case DELETE_FAST:
-                unsafe = true;
+            case STORE_FAST:
+                unsafe_mask &= ~bit;
                 break;
         }
     }
-    if (unsafe) {
-        // unsafe at end of this block,
-        // so unsafe at start of next blocks
-        if (b->b_next && BB_HAS_FALLTHROUGH(b)) {
-            MAYBE_PUSH(b->b_next);
-        }
-        struct instr *last = basicblock_last_instr(b);
-        if (last != NULL) {
-            if (is_jump(last)) {
-                assert(last->i_target != NULL);
-                MAYBE_PUSH(last->i_target);
-            }
-        }
+    if (b->b_next && BB_HAS_FALLTHROUGH(b)) {
+        maybe_push(b->b_next, unsafe_mask, sp);
+    }
+    struct instr *last = basicblock_last_instr(b);
+    if (last && is_jump(last)) {
+        assert(last->i_target != NULL);
+        maybe_push(last->i_target, unsafe_mask, sp);
     }
 }
-#undef MAYBE_PUSH
 
 static int
 add_checks_for_loads_of_unknown_variables(basicblock *entryblock,
                                           struct compiler *c)
 {
+    int nparams = (int)PyList_GET_SIZE(c->u->u_ste->ste_varnames);
+    int nlocals = (int)PyDict_GET_SIZE(c->u->u_varnames);
+    if (nlocals > 64) {
+        // Avoid O(nlocals**2) compilation:
+        // only analyze the first 64 locals.
+        // The rest get only LOAD_FAST_CHECK.
```

Comment:
It would possible to allocate `nlocals` bytes or bits and carry through the "certainly initialized" information within each basicblock, so that any contiguous `a9999 = 1; print(a9999)` would use `LOAD_FAST` (without check), no matter how many variables are present in the function. This would retain the linear time as well, but not propagate information between basicblocks.

It's unclear to me whether that is worth the few more lines of code and logic.

Reply:
Probably, we need to benchmark both variants for 63 and 9999 variables first to drop the idea if it caused a performance drop for 63 or gave a negligible gain for 9999.